Add CRUD endpoints for dataset metadata

nclskfm opened the pull request:
## What does this PR do?
#7 

* Add CRUD endpoints
* Add auth
* Update DatasetFormatter to use the dataset mappings
* Remove mocked `get_dataset_mapping` function
* Add DatasetHandler to download/delete datasets as background task
* Add comments
* Add logging

## Todo:

- [x] Add logging
- [x] Add comments
- [x] Replace mocked `get_dataset_mapping` function
- [x] Add auth check for admins
- [x] Call Datasethandler

BenediktStuhrmann commented on evaluator/evaluator/tasks/predict_task.py:
```diff
@@ -99,20 +99,24 @@ def do_predict(
             query_request = {
-                "query": reference_data["question"],
-                "skill_args": {"context": reference_data["context"]},
+                "query": reference_data[dataset_metadata.mapping.question_column],
```
Die `query_request` Sachen musst du wieder zurückändern. "question", "context", "choices" beziehen sich an der Stelle auf die entsprechenden Felder von `ExtractiveQASample ` und `MultipleChoiceSample`. Nicht auf das Mapping aus den Metadaten! (Multiple-choice funktioniert aktuell z.B. so garnicht (wenn ich es zurückändere dann aber schon)

nclskfm replied:
Ups, ja konnte vorhin das nicht ausprobieren, weil square wieder down war und vor paar Tagen hatte ich dann vermutlich nur ExtractiveQA getestet...

BenediktStuhrmann commented on evaluator/evaluator/app/models.py:
```diff
@@ -9,24 +9,42 @@
+    id_column: str = Field(...)
+    question_column: str = Field(...)
+    choices_columns: list[str] = Field(...)
+    choices_key_mapping_column: str = Field(...)
```
Das muss optional sein. z.B.
```json
{
  "name": "commonsense_qa",
  "skill_type": "multiple-choice",
  "metric": "exact_match",
  "mapping": {
    "id_column": "id",
    "question_column": "question",
    "choices_columns": ["choices.text"],
    "choices_key_mapping_column": "choices.label",
    "answer_index_column": "answerKey"
  }
}
```

vs. 

```json
{
  "name": "cosmos_qa",
  "skill_type": "multiple-choice",
  "metric": "exact_match",
  "mapping": {
    "id_column": "id",
    "question_column": "question",
    "choices_columns": ["answer0", "answer1", "answer2", "answer3"],
    "choices_key_mapping_column": null,
    "answer_index_column": "label"
  }
}
```

BenediktStuhrmann commented on evaluator/evaluator/app/routers/dataset.py:
```diff
@@ -16,7 +26,185 @@
+    validate_access(token_payload)
+    await validate_mapping_and_skill_type(dataset_metadata)
+    validate_metric(dataset_metadata.metric)
+    validate_dataset_name(dataset_metadata.name)
```
Ich würde den Namen tatsächlich immer als erstes (nach der Zugriffskontrolle) validieren. Gleiches dann auch beim PUT

BenediktStuhrmann commented on evaluator/evaluator/app/routers/dataset.py:
```diff
@@ -16,7 +26,185 @@
-    datasets = [dataset.value for dataset in DataSet]
 
+    results = mongo_client.client.evaluator.datasets.find()
+    datasets = [DatasetMetadata.from_mongo(result).name for result in results]
```
Wieso returnen wir nicht die Liste von datasets? Sondern nur die Namen 🤔 Für das Frontend (Leadeboard) bräuchte ich die gesamten Datasets (also alle Metadaten, nicht nur den Namen jeweils).

nclskfm replied:
Oh, dachte, das braucht man nicht. Dann ändere ich das :)

BenediktStuhrmann commented on evaluator/evaluator/app/routers/dataset.py:
```diff
@@ -16,7 +26,185 @@
+    """
+    Function to validate access. Currently SQuARE has no user roles, so we check for the statically defined user names.
+    """
+    ALLOWED_USER_NAMES = ["ukp", "LOCAL_SQUARE_USER"]
```
Würde die lieber kleinschreiben und dann hier `if token_payload["username"] not in ALLOWED_USER_NAMES:` den username aus dem Token lowercasen. Nicht, dass es am Ende scheitert weil ukp groß geschrieben wird oder sowas :D

nclskfm replied:
Wollte das als extra Sicherheitsfeature, aber sicher ist das wahrscheinlich eh nicht 😂 Dann ändere ich das

BenediktStuhrmann commented on evaluator/evaluator/app/routers/dataset.py:
```diff
@@ -16,7 +26,185 @@
+
+
+def trigger_download_dataset(dataset_handler: DatasetHandler, dataset_name: str):
+    dataset_handler.download_dataset(dataset_name)
```
Wenn ich das richtig sehe, wird jetzt bei jedem Update vom z.B. Mapping des Datasets auch das dataset neu runtergeladen (auch wenn es schon existiert lokal). Würde das also in `dataset_handler.get_dataset()` abändern.

Außer man sagst, das ist ggf. ein Feature?

nclskfm replied:
Joa, dachte das ist ein Feature, weil sich die Datasetdaten ja ggf. ändern können (z.B. wenn mehr Datenpunkte hinzugefügt werden). Deswegen lieber immer Neuladen forcen

BenediktStuhrmann reviewed (changes requested):
Sieht gut aus. Hatte es auch lokal getestet mal. 🚀 
Bei multiple-choice scheitert es gerade wegen dem `query_request` im predict-task (siehe entsprechender Kommentar).
Und dann noch eine kleinigkeiten beim multiple-choice-mapping.

Sonst scheint alles zu klappen 👍🏻

nclskfm commented:
Code suggestions / Bug fixes:

* choices_key_mapping_column is optional
* GET /datasets return whole dataset objects instead of the names only
* Moved `validate_dataset_name`
* Changed access validation to be case-insensitive
* Fix predict task

BenediktStuhrmann reviewed (approved):
Sieht gut aus 👍🏻
